include/RAJA/exec-agency/forallN_agency.hxx

```diff
@@ -0,0 +1,211 @@
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/int_datatypes.hxx"
+
+#include "agency/agency.hpp"
+#include "raja_agency.hxx"
+
+namespace RAJA
+{
+
+/******************************************************************
+ *  ForallN policies
+ ******************************************************************/
+
+template <typename Agent, typename Worker, typename... PREST>
+struct ForallN_Executor<ForallN_PolicyPair<agency_base<Agent, Worker>,
+                                           RangeSegment>,
+                        ForallN_PolicyPair<agency_base<Agent, Worker>,
+                                           RangeSegment>,
+                        PREST...> {
+  using PolicyPair = ForallN_PolicyPair<agency_base<Agent, Worker>, RangeSegment>;
+  PolicyPair iset_i, iset_j;
+
+  using NextExec = ForallN_Executor<PREST...>l
+  NextExec next_exec;
+
+  RAJA_INLINE
+  constexpr ForallN_Executor(
+      PolicyPair const &iseti_, 
+      PolicyPair const &isetj_,
+      PREST const &... prest)
+      : iset_i(iseti_), iset_j(isetj_), next_exec(prest...)
+  {
+  }
+
+  template <typename BODY>
+  RAJA_INLINE void operator()(BODY body) const
+  {
+    int begin_i = iset_i.getBegin();
+    int begin_j = iset_j.getBegin();
+    int end_i = iset_i.getEnd();
+    int end_j = iset_j.getEnd();
+
+    ForallN_PeelOuter<NextExec, BODY> outer(next_exec, body);
+
+    auto numThreads = max(std::thread::hardware_concurrency(), 1);
+    auto workPerThread = (end_i - begin_i) / numThreads;
+
+    agency::bulk_invoke(Worker(numThreads),
+                        [=](Agent& self) {
+                          auto start = begin_i + workPerThread * self.index();
+                          auto end = self.index() == (numThreads - 1)
+                                        ? end_i
+                                        : start + workPerThread;
+                          for (auto i = start; i < end; ++i) {              
+                            for (auto j = begin_j; j < end_j; ++j) {
+                              outer(i, j);
+                            }
+                          }
+                        });
+  }
+};
+
+template <typename Agent, typename Worker, typename... PREST>
+struct ForallN_Executor<ForallN_PolicyPair<agency_base<Agent, Worker>,
+                                           RangeSegment>,
+                        ForallN_PolicyPair<agency_base<Agent, Worker>,
+                                           RangeSegment>,
+                        ForallN_PolicyPair<agency_base<Agent, Worker>,
+                                           RangeSegment>,
+                        PREST...> {
+  using PolicyPair = ForallN_PolicyPair<agency_base<Agent, Worker>, RangeSegment>;
+  PolicyPair iset_i, iset_j, iset_k;
+
+  using NextExec = ForallN_Executor<PREST...>l
+  NextExec next_exec;
+
+  RAJA_INLINE
+  constexpr ForallN_Executor(
```

Comment:
I'm not totally clear on it either, however its the same way that the openMP forallN is written. I can ask Holger/David.